### PR TITLE
feat: support AP Group resource

### LINKED ISSUE #81: Support for `unifi_ap_group` resource
# **🚀 Feature Request**

## **Is your feature request related to a problem? Please describe.**

* I want to manage(create / update / delete) AP Group in Terraform.

---

## **Describe the solution you'd like**

* Create a resource type: `unifi_ap_group`

---

## **Describe alternatives you've considered**

*

---

### **Additional context**

* There is only data source for AP Group.

### PR DESCRIPTION
**Pull request by filipowm**
Closes #81

### CODE REVIEW

**Review by filipowm — COMMENTED** [review]
-

File: internal/provider/apgroup/resource_ap_group.go [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+					stringvalidator.LengthAtLeast(1),
+				},
+			},
+			"device_macs": schema.ListAttribute{
+				MarkdownDescription: "List of AP devices MAC addresses to include in this AP group.",
+				Required:            true,
+				ElementType:         types.StringType,
+				Validators: []validator.List{
+					listvalidator.SizeAtLeast(1),
+					listvalidator.ValueStringsAre(validators.Mac),
```
Comment (filipowm) [inline internal/provider/apgroup/resource_ap_group.go]: 🟠 **device_macs not normalized to controller's canonical MAC form** — `high` · correctness

device_macs is a Required (non-computed) attribute whose validator now accepts uppercase and hyphen-separated MACs, but AsUnifiModel/Merge pass the values through unchanged. The UniFi controller stores MACs lowercase and colon-separated, so a config like ["AA:BB:CC:DD:EE:FF"] (or dash form) round-trips to a different string, producing a "Provider produced inconsistent result after apply" error that fails the apply (leaving the AP group created on the controller but absent from state), and otherwise a perpetual diff.

**Fix — options:**
- **Semantic-equality custom type** — Define a MAC-typed string (StringSemanticEquals via CleanMAC) and use it for device_macs elements so Terraform treats separator/case variants as equal — no inconsistent-apply error.
- **Plan modifier normalization** — Add a list/string plan modifier that rewrites device_macs to CleanMAC (lowercase colon) form during plan so the planned value equals the controller's canonical response.
- **Tighten validator** — Restrict device_macs to canonical lowercase, colon-separated form so the config value already equals what the controller stores (rejects uppercase/dash up front).

**Recommended:** Make config-side MACs canonicalize at plan time so the planned value already matches what the controller returns — mirroring the legacy utils.CleanMAC + MacDiffSuppressFunc behavior. Normalizing only inside AsUnifiModel is insufficient because the planned (config) value still differs from the post-apply state. Prefer a reusable custom MAC string type implementing StringSemanticEquals (or a string plan modifier) applied to device_macs elements; alternatively tighten the validator to canonical lowercase-colon form so non-canonical input is rejected up front.

<details>
<summary>Details</summary>

Triggering schedule:
1. User config: device_macs = ["00-15-6D-00-00-01"] (dash + uppercase) — accepted by the widened validators.MacRegex `^([0-9a-fA-F][0-9a-fA-F][-:]){5}([0-9a-fA-F][0-9a-fA-F])$` and by listvalidator.ValueStringsAre(validators.Mac) (resource_ap_group.go:121).
2. Create: AsUnifiModel (resource_ap_group.go:27-41) copies the list verbatim into unifi.APGroup.DeviceMACs and calls client.CreateAPGroup.
3. The controller canonicalizes the MAC to 00:15:6d:00:00:01 and returns it in the create response.
4. base/generic_resource.go:121 plan.Merge(ctx, res) → APGroupModel.Merge (resource_ap_group.go:56) overwrites DeviceMACs in state with ["00:15:6d:00:00:01"] via types.ListValueFrom.
5. resp.State.Set writes state. Terraform compares the post-apply state value against the planned value ["00-15-6D-00-00-01"] for this Required (non-computed) attribute → mismatch → 'Provider produced inconsistent result after apply'. The apply errors AFTER the controller object was created, so the AP group exists on the controller but is not tracked in state (manual import to recover); on the Read path it reproduces as a perpetual diff.

This is exactly the case the legacy SDKv2 paths guard against via utils.CleanMAC + MacDiffSuppressFunc (used by device/user/wlan resources; resource_user.go even documents "MAC addresses are case-insensitive"). The new Framework resource has no equivalent semantic-equality type or plan-time normalization, and no ModifyPlan/PlanModifier exists in the package. The same outcome occurs for plain uppercase-colon input 00:15:6D:00:00:01, so the gap is a property of the new resource; the MacRegex widening in this same PR (validators/regex.go:19) merely enlarges the set of inputs that hit it. The project's own (skipped) TestAccAPGroup_update uses mac2 = "AA:BB:CC:DD:EE:FF", which would trip this — but TestAccAPGroup_basic/_update are t.Skip-ped, so the create/update round-trip is never exercised.

</details>

<sub>lens: architect, Reliability & Concurrency, tf-dev · confidence: high</sub>

File: internal/provider/apgroup/resource_ap_group.go [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+			"id":   ut.ID(),
+			"site": ut.SiteAttribute(),
+			"name": schema.StringAttribute{
+				MarkdownDescription: "The name of the AP group.",
+				Required:            true,
+				Validators: []validator.String{
+					stringvalidator.LengthAtLeast(1),
+				},
+			},
+			"device_macs": schema.ListAttribute{
```
Comment (filipowm) [inline internal/provider/apgroup/resource_ap_group.go]: 🔵 **device_macs modeled as ordered List, not Set** — `low` · design

AP-group membership is an unordered set, but device_macs is a schema.ListAttribute (types.List), so any reordering of the MACs by the controller or the user surfaces as a spurious diff with no semantic meaning.

**Fix:** Use schema.SetAttribute with types.Set for device_macs (and the model field) so membership is compared order-insensitively. Note this touches the model field, Merge (ListValueFrom->SetValueFrom), validator imports (listvalidator->setvalidator), and AsUnifiModel — not just the schema line.

<sub>lens: architect · confidence: high</sub>

File: internal/provider/apgroup/data_ap_group.go [hunk 3]
```diff
@@ -2,58 +2,133 @@ package apgroup
...
+type APGroupDatasourceModel struct {
+	base.Model
+	Name types.String `tfsdk:"name"`
+}
 
-		ReadContext: DataAPGroupRead,
+// AsUnifiModel converts the Terraform model to the UniFi API model
+func (m *APGroupDatasourceModel) AsUnifiModel(_ context.Context) (interface{}, diag.Diagnostics) {
+	return nil, diag.Diagnostics{}
+}
```
Comment (filipowm) [inline internal/provider/apgroup/data_ap_group.go:25]: 🔵 **Unused AsUnifiModel no-op on datasource model** — `low` · dead-code

APGroupDatasourceModel.AsUnifiModel returns nil and is never called. The base.DatasourceModel interface only requires SiteAware + Merge, the model is never asserted as base.ResourceModel, and the Read path (Get/ResolveSite/Merge/SetID/SetSite/State.Set) never invokes it, so this no-op is dead code that misleadingly implies the data source writes back to the API.

**Fix:** Delete the AsUnifiModel method (and its doc comment) from APGroupDatasourceModel; the data source never converts model→API.

```suggestion

```

> _Removes the unused no-op method and its trailing blank line; context/diag imports remain used by Merge._

<details>
<summary>Details</summary>

Unlike dns, where dnsRecordModel is shared between the resource and data source (so its AsUnifiModel is exercised by the resource), apgroup has a separate APGroupModel for the resource. APGroupDatasourceModel exists only for the read-only data source. The only AsUnifiModel call sites are base/generic_resource.go (on GenericResource[T ResourceModel] plan models), and apgroup uses a hand-written apGroupDatasource, not GenericResource. dnsRecordsDatasourceModel also defines no AsUnifiModel, confirming the pattern. Deleting the method compiles cleanly and leaves context/diag imports in use via Merge.

</details>

<sub>lens: architect, readability · confidence: high</sub>

File: internal/provider/acctest/resource_ap_group_test.go [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+	"regexp"
+	"strings"
+	"testing"
+)
+
+const testAPGroupResourceName = "unifi_ap_group.test"
+
+// FIXME: This test is currently skipped because the test environment does not support AP group adoption.
+func TestAccAPGroup_basic(t *testing.T) {
+	t.Skip("Skipping test due to lack of support for AP group adoption in test environment")
```
Comment (filipowm) [inline internal/provider/acctest/resource_ap_group_test.go:20]: 🔵 **AP group resource CRUD/import lifecycle entirely skipped** — `low` · testing

The only two behavioral tests for the new resource, TestAccAPGroup_basic (line 20) and TestAccAPGroup_update (line 43), both call t.Skip, so Create/Read/Update/Delete, import round-trip, plan stability, and CheckDestroy are never executed. The resource ships with zero exercised happy-path behavior; only schema-validation error paths run.

**Fix:** Add a controller-free unit test for APGroupModel.AsUnifiModel and Merge (round-trip of ID, Name, and device_macs) so the model conversion is pinned even while the acceptance lifecycle tests remain skipped; revisit un-skipping basic/update once the environment supports AP-group adoption.

<details>
<summary>Details</summary>

Consequence: regressions in AsUnifiModel/Merge (model<->API conversion), the device_macs list round-trip (types.ListValueFrom), in-place update vs ForceNew, or perpetual-diff stability would not be caught by CI. The new pt.TestCheckListResourceAttr helper and testAccCheckAPGroupDestroy are referenced only by these skipped tests, so they are effectively dead code today. The skip reason (test environment lacks AP-group adoption) is legitimate for the full apply path, but it leaves the model-conversion logic completely unpinned. A controller-free unit test in the apgroup package over a *unifi.APGroup fixture asserting AsUnifiModel(model).DeviceMACs and a Merge(...) round-trip (including ID/Name/device_macs) would pin the conversion the skipped acceptance tests were meant to cover, without needing Docker.

</details>

<sub>lens: testing · confidence: medium</sub>

File: internal/provider/validators/regex.go [hunk 5]
```diff
@@ -16,7 +16,7 @@ var (
 	PortRangeRegexp = regexp.MustCompile("(([1-9][0-9]{0,3}|[1-5][0-9]{4}|[6][0-4][0-9]{3}|[6][5][0-4][0-9]{2}|[6][5][5][0-2][0-9]|[6][5][5][3][0-5])|([1-9][0-9]{0,3}|[1-5][0-9]{4}|[6][0-4][0-9]{3}|[6][5][0-4][0-9]{2}|[6][5][5][0-2][0-9]|[6][5][5][3][0-5])-([1-9][0-9]{0,3}|[1-5][0-9]{4}|[6][0-4][0-9]{3}|[6][5][0-4][0-9]{2}|[6][5][5][0-2][0-9]|[6][5][5][3][0-5]))+(,([1-9][0-9]{0,3}|[1-5][0-9]{4}|[6][0-4][0-9]{3}|[6][5][0-4][0-9]{2}|[6][5][5][0-2][0-9]|[6][5][5][3][0-5])|,([1-9][0-9]{0,3}|[1-5][0-9]{4}|[6][0-4][0-9]{3}|[6][5][0-4][0-9]{2}|[6][5][5][0-2][0-9]|[6][5][5][3][0-5])-([1-9][0-9]{0,3}|[1-5][0-9]{4}|[6][0-4][0-9]{3}|[6][5][0-4][0-9]{2}|[6][5][5][0-2][0-9]|[6][5][5][3][0-5])){0,14}")
 	PortRangeV2     = validation.StringMatch(PortRangeRegexp, "invalid port range")
 
-	MacRegex = regexp.MustCompile(`^([0-9A-Fa-f]{2}:){5}([0-9A-Fa-f]{2})$`)
+	MacRegex = regexp.MustCompile(`^([0-9a-fA-F][0-9a-fA-F][-:]){5}([0-9a-fA-F][0-9a-fA-F])$`)
```
Comment (filipowm) [inline internal/provider/validators/regex.go:19]: 🔵 **Broadened MAC regex has no positive-case or unit coverage** — `low` · testing

MacRegex was widened to accept '-' separators (and, as a side effect, mixed separators like 00-11:22:33-44:55), but the only test touching it is TestAccAPGroup_invalidMacAddress, which checks a single rejected input. No test asserts a dash-separated MAC is now accepted, and regex.go has no _test.go even though every sibling validator (cidr, ipv4, hostname, etc.) does.

**Fix:** Add validators/regex_test.go with a table-driven test over MacRegex covering accepted colon and dash forms plus rejected garbage/length/format cases; explicitly decide whether mixed-separator MACs should be accepted and assert that decision.

<details>
<summary>Details</summary>

The accepted-MAC inputs only appear in TestAccAPGroup_basic/_update, both of which are skipped, so no test ever exercises an accepted MAC at all (dash or colon). A regression that broke dash acceptance, re-narrowed to colon-only, or further broadened the pattern would pass CI. Because validation runs at plan time, this can be pinned cheaply with a table-driven Go unit test (validators/regex_test.go) over MacRegex.MatchString with accepted cases ('00:11:22:33:44:55', '00-11-22-33-44-55') and rejected cases ('invalid-mac-address', '0011.2233.4455', too-short/too-long). Such a test would also surface the mixed-separator acceptance for an explicit decision rather than leaving it silent and untested.

</details>

<sub>lens: testing · confidence: high</sub>

File: internal/provider/acctest/data_ap_group_test.go [hunk 6]
```diff
@@ -1,24 +1,51 @@
...
+
 func TestAccDataAPGroup_default(t *testing.T) {
 	AcceptanceTest(t, AcceptanceTestCase{
 		Steps: []resource.TestStep{
 			{
 				Config: testAccDataAPGroupConfig_default,
-				Check:  resource.ComposeTestCheckFunc(
-				// testCheckNetworkExists(t, "name"),
+				Check: resource.ComposeTestCheckFunc(
+					resource.TestCheckResourceAttr(testAPGroupDatasourceName, "name", defaultAPGroupName),
```
Comment (filipowm) [inline internal/provider/acctest/data_ap_group_test.go:20]: 🔵 **AP group data source tests assert only name, not the id** — `low` · testing

TestAccDataAPGroup_default (line 20) and TestAccDataAPGroup_byName (line 33) both assert only the 'name' attribute, but the data source exists to 'retrieve the ID for an AP group by name'. A regression in SetID(found.ID) (data_ap_group.go:131) or in site resolution would still pass, and the not-found error branch (groups empty / no name match) is never exercised.

**Fix:** Add resource.TestCheckResourceAttrSet(testAPGroupDatasourceName, "id") (and a "site" check) to both data-source checks, and add a case for a non-existent name asserting ExpectError on the 'AP group not found' diagnostic to cover the miss branch.

<sub>lens: testing · confidence: high</sub>

File: internal/provider/apgroup/resource_ap_group.go [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+				Delete: func(ctx context.Context, client *base.Client, site, id string) error {
+					return client.DeleteAPGroup(ctx, site, id)
+				},
+			},
+		),
+	}
+}
+
+// Schema defines the schema for the resource
+func (r *apGroupResource) Schema(_ context.Context, _ resource.SchemaRequest, resp *resource.SchemaResponse) {
```
Comment (filipowm) [inline internal/provider/apgroup/resource_ap_group.go:109]: 🔵 **New unifi_ap_group resource ships without example files and generated docs** — `low` · documentation

The new, importable unifi_ap_group resource is registered in provider_v2.go but has no examples/resources/unifi_ap_group/resource.tf or import.sh, and no docs/resources/ap_group.md was generated. The project's docs are generated by tfplugindocs from each schema plus the examples/ tree, and the project rule requires examples/resources/<name>/resource.tf (plus import.sh for importable resources) when a resource is introduced.

**Fix — options:**
- **Add resource.tf** — resource "unifi_ap_group" "example" { name = "my-ap-group"  device_macs = ["00:11:22:33:44:55"] }
- **Add import.sh** — terraform import unifi_ap_group.example 5dc28e5e9106d105bdc87217 (and the another-site:ID variant), matching examples/resources/unifi_firewall_zone/import.sh

**Recommended:** Add examples/resources/unifi_ap_group/resource.tf (a unifi_ap_group block with name and device_macs) and examples/resources/unifi_ap_group/import.sh (following the existing examples/resources/unifi_firewall_zone/import.sh site:id pattern), then run `go generate ./tools/` to produce docs/resources/ap_group.md.

<details>
<summary>Details</summary>

The data source already has examples/data-sources/unifi_ap_group/data-source.tf, but no examples/resources/unifi_ap_group/ directory exists and no docs/resources/ap_group.md is produced. The resource asserts resource.ResourceWithImportState (import via base.ImportIDWithSite, site:id format, line 66) and the acceptance test uses pt.ImportStepWithSite, so an import.sh is warranted. (The stale data-source docs page is tracked separately in F-008.)

</details>

<sub>lens: documentation, tf-dev · confidence: high</sub>

File: internal/provider/apgroup/data_ap_group.go [hunk 8]
```diff
@@ -2,58 +2,133 @@ package apgroup
...
+	base.ConfigureDatasource(d, req, resp)
+}
+
+func (d *apGroupDatasource) Metadata(_ context.Context, _ datasource.MetadataRequest, resp *datasource.MetadataResponse) {
+	resp.TypeName = "unifi_ap_group"
+}
+
+func (d *apGroupDatasource) Schema(_ context.Context, _ datasource.SchemaRequest, resp *datasource.SchemaResponse) {
+	resp.Schema = schema.Schema{
+		MarkdownDescription: "The `unifi_ap_group` data source can be used to retrieve the ID for an AP group by name.",
```
Comment (filipowm) [inline internal/provider/apgroup/data_ap_group.go:80]: 🔵 **Data-source description changes not regenerated into docs** — `low` · documentation

The Framework rewrite changes the data source's descriptions (top-level now prefixed with "The ", id via ut.ID() to "The unique identifier of this resource", and site via ut.SiteAttribute()), but committed docs/data-sources/ap_group.md still shows the old SDKv2 text, so Registry docs are stale.

**Fix:** Run `go generate ./tools/` to refresh docs/data-sources/ap_group.md so the rendered id/site/description text matches the new Framework schema.

<sub>lens: tf-dev · confidence: high</sub>

File: internal/provider/apgroup/data_ap_group.go [hunk 9]
```diff
@@ -2,58 +2,133 @@ package apgroup
...
+		Attributes: map[string]schema.Attribute{
+			"id":   ut.ID(),
+			"site": ut.SiteAttribute(),
+			"name": schema.StringAttribute{
+				MarkdownDescription: "The name of the AP group to look up, leave blank to look up the default AP group.",
+				Optional:            true,
+				Validators: []validator.String{
+					stringvalidator.LengthAtLeast(1),
```
Comment (filipowm) [inline internal/provider/apgroup/data_ap_group.go]: 🔵 **name="" rejected despite "leave blank" documentation** — `low` · correctness

The data source's name attribute documents "leave blank to look up the default AP group," but the newly added stringvalidator.LengthAtLeast(1) rejects an explicit name = "" with a length error instead of resolving the default group. Only omitting the attribute (null) reaches the default-lookup branch.

**Fix — options:**
- **Drop the validator** — Remove LengthAtLeast(1) so name = "" falls through to the existing default-group lookup, matching the description.
- **Reword the description** — Change the description to instruct omitting the attribute (rather than "leave blank") to look up the default group.

**Recommended:** Pick one so behavior and docs agree; removing the validator restores the documented name="" default lookup with the least surprise.

<sub>lens: tf-dev · confidence: high</sub>

**Review by filipowm — COMMENTED** [review]
## PR Review

This PR adds a first-class `unifi_ap_group` resource and migrates the pre-existing `unifi_ap_group` data source from the legacy SDKv2 to the Terraform Plugin Framework. Both are now registered in the Framework provider (`provider_v2.go`) and removed from the SDKv2 provider (`provider.go`), satisfying the project's mux-placement rule that a given name live in exactly one provider. The new resource is built on `base.GenericResource[*APGroupModel]` with CRUD wired to the go-unifi AP-group client calls, exposes a required `name` and a required `device_macs` list, and supports `site:id` import. Supporting changes widen the shared `MacRegex` validator to also accept hyphen-separated MACs, add a `TestCheckListResourceAttr` test helper, and add acceptance tests — though the resource's create/update/import lifecycle tests are hard-skipped because the Dockerized controller cannot adopt APs, so only schema-validation error paths are actually exercised.

### Walkthrough

**New unifi_ap_group resource and provider registration**

Adds resource_ap_group.go: an APGroupModel (with AsUnifiModel/Merge round-trip) backed by base.GenericResource[*APGroupModel], a schema exposing required name and a required device_macs list (size>=1, each MAC-validated), and ResourceWithImportState. Registers NewAPGroupResource and NewAPGroupDatasource in the Framework provider, and removes the old SDKv2 data-source registration plus the resource TODO from the legacy provider so each name lives in exactly one provider.

Touches: `internal/provider/apgroup/resource_ap_group.go` L1-L126, `internal/provider/provider_v2.go` L179-L205, `internal/provider/provider.go` L95-L103

**AP Group data source rewritten on the Plugin Framework**

Replaces the SDKv2 DataAPGroup/DataAPGroupRead with a Framework datasource.DataSource: APGroupDatasourceModel plus Configure/Metadata/Schema/Read, reusing ut.ID() and ut.SiteAttribute(), resolving the site via client.ResolveSite, emitting structured diagnostics, and adding an explicit empty/not-found branch. Adds a stringvalidator.LengthAtLeast(1) on name and an unused AsUnifiModel no-op on the datasource model.

Touches: `internal/provider/apgroup/data_ap_group.go` L23-L31, `internal/provider/apgroup/data_ap_group.go` L82-L95, `internal/provider/apgroup/data_ap_group.go` L119-L132

**Widened MAC validator and list-attribute test helper**

Broadens MacRegex from colon-only to accept '-' as well as ':' separators (and, as a side effect, mixed separators like 00-11:22:33-44:55), and adds a reusable TestCheckListResourceAttr helper that asserts a list attribute's count and per-index element values.

Touches: `internal/provider/validators/regex.go` L19, `internal/provider/testing/test_helpers_objects.go` L1-L15

**AP Group acceptance tests**

Adds resource_ap_group_test.go with basic and update lifecycle tests (both t.Skip-ped due to missing AP-group adoption in the test env), validation-error tests (missing name, missing/empty device_macs, invalid MAC), and a CheckDestroy. Extends data_ap_group_test.go with default and by-name lookup tests, each asserting only the name attribute.

Touches: `internal/provider/acctest/resource_ap_group_test.go` L1-L175, `internal/provider/acctest/data_ap_group_test.go` L1-L58

### Diagrams

#### Why a non-canonical device_macs value fails apply (F-001)

```mermaid
sequenceDiagram
    participant TF as Terraform Core
    participant R as apGroupResource
    participant C as UniFi Controller
    TF->>R: Create(plan device_macs=["00-15-6D-00-00-01"])
    R->>R: AsUnifiModel (copied verbatim, no normalization)
    R->>C: CreateAPGroup(DeviceMACs)
    C-->>R: APGroup{DeviceMACs:["00:15:6d:00:00:01"]}
    R->>R: Merge -> state device_macs=["00:15:6d:00:00:01"]
    R-->>TF: State.Set
    TF->>TF: compare planned vs post-apply value
    Note over TF: mismatch on a Required attr -> "Provider produced inconsistent result after apply"
```

### SemVer impact

**minor** — The strongest signal is a backward-compatible public addition: a brand-new unifi_ap_group resource is registered in the Framework provider, leaving existing consumers working. The unifi_ap_group data source is reimplemented on the Plugin Framework but keeps its name and id/site/name attributes, and the MacRegex change only widens the set of accepted inputs; the only behavioral regression (name="" now rejected, F-009) is a minor, low-severity edge case that does not rise to a forced major bump.

### Verdict

> [!CAUTION]
> **Request changes** — The new resource is well-structured and correctly placed in the Framework provider, but device_macs is not normalized to the controller's canonical MAC form (F-001, high), so any accepted non-canonical input round-trips to a different string and fails the apply with an inconsistent-result error. Compounding this, the resource's create/update/import lifecycle is entirely t.Skip-ped, so no test exercises the conversion that would have caught it. Address the MAC normalization gap (and ideally pin the model conversion with a controller-free unit test) before merge.

### Findings

| Severity | Count |
| --- | --- |
| Critical | 0 |
| High | 1 |
| Medium | 0 |
| Low | 8 |
| Info | 0 |
| **Total** | **9** |

- 🟠 **F-001** `high` — device_macs not normalized to controller's canonical MAC form (`internal/provider/apgroup/resource_ap_group.go` L115-L121)
- 🔵 **F-002** `low` — device_macs modeled as ordered List, not Set (`internal/provider/apgroup/resource_ap_group.go` L115)
- 🔵 **F-003** `low` — Unused AsUnifiModel no-op on datasource model (`internal/provider/apgroup/data_ap_group.go` L23-L27)
- 🔵 **F-004** `low` — AP group resource CRUD/import lifecycle entirely skipped (`internal/provider/acctest/resource_ap_group_test.go` L19-L20)
- 🔵 **F-005** `low` — Broadened MAC regex has no positive-case or unit coverage (`internal/provider/validators/regex.go` L19)
- 🔵 **F-006** `low` — AP group data source tests assert only name, not the id (`internal/provider/acctest/data_ap_group_test.go` L19-L20)
- 🔵 **F-007** `low` — New unifi_ap_group resource ships without example files and generated docs (`internal/provider/apgroup/resource_ap_group.go` L99)
- 🔵 **F-008** `low` — Data-source description changes not regenerated into docs (`internal/provider/apgroup/data_ap_group.go` L82)
- 🔵 **F-009** `low` — name="" rejected despite "leave blank" documentation (`internal/provider/apgroup/data_ap_group.go` L86-L90)

<sub>Automated review — verify findings before acting.</sub>

**Comment by filipowm** [comment]
<details><summary>pr-review incremental state — do not edit</summary>

This hidden marker records the last-reviewed commit and last-seen comment id so that
`--incremental` re-reviews only what changed. It is managed automatically by the pr-review skill.

</details>

File: internal/provider/acctest/resource_ap_group_test.go [hunk 10]
```diff
@@ -0,0 +1,175 @@
...
+		}
+
+		_, err := testClient.GetAPGroup(context.Background(), "default", rs.Primary.ID)
+		if err == nil {
+			return fmt.Errorf("AP Group %s still exists", rs.Primary.ID)
+		}
+
+		// If we get a 404 error, that means the resource was deleted
+		if strings.Contains(err.Error(), "404") || strings.Contains(err.Error(), "not found") {
+			continue
```
[the remark this answers is not in the document]
Comment (filipowm) [reply internal/provider/acctest/resource_ap_group_test.go]: Good catch — `testAccCheckAPGroupDestroy` hard-codes `"default"`, so a test running against any other site would query the wrong place and could mis-report destroy. Fixed by reading the site off the destroyed resource's state and falling back to `"default"`:

```go
site := rs.Primary.Attributes["site"]
if site == "" {
    site = "default"
}

_, err := testClient.GetAPGroup(context.Background(), site, rs.Primary.ID)
```

The rest of the loop (the `err == nil` "still exists" check and the 404/not-found continue) is unchanged. `make build` passes; the `acctest` package compiles and vets clean, gofmt reports no diff, and the targeted `TestAccAPGroup` test binary builds and runs. Note: the full `make testacc` run was not exercised in this environment because Docker (required to start the Dockerized controller) was unavailable.

File: internal/provider/apgroup/data_ap_group.go [hunk 11]
```diff
@@ -2,58 +2,133 @@ package apgroup
...
+				},
```
[the remark this answers is not in the document]
Comment (filipowm) [reply internal/provider/apgroup/data_ap_group.go]: Correct — the `name` description says "leave blank to look up the default AP group," but `stringvalidator.LengthAtLeast(1)` rejected an explicit `name = ""` with a length error before the Read path could reach the default-group branch (only omitting the attribute / `null` worked). Resolved by dropping the validator so an empty string falls through to the existing `name == "" && g.HiddenID == "default"` lookup, matching the documented behavior.

Scope is data-source only: the matching `name` attribute in `resource_ap_group.go` is `Required` and keeps its `LengthAtLeast(1)`. The now-unused `stringvalidator` and `schema/validator` imports were removed.

Verification: `make build` (go install), `go build ./...`, and `golangci-lint run --fix ./...` (golangci-lint v1.64.8, matching CI) all pass with no findings; `go vet` is clean on both the `apgroup` and `acctest` packages. Generated docs are unaffected — only the (unchanged) description is rendered. `make testacc` could not run in this environment because no Docker daemon is available to start the controller; the change is compile-clean and behavioral only.

File: internal/provider/apgroup/resource_ap_group.go [hunk 12]
```diff
@@ -0,0 +1,126 @@
...
+	resp.Schema = schema.Schema{
+		MarkdownDescription: "The `unifi_ap_group` resource manages Access Point groups in the UniFi controller.\n\n" +
+				"AP groups allow you to organize and manage multiple access points together. " +
+				"This resource allows you to create, update, and delete AP groups.",
```
[the remark this answers is not in the document]
Comment (filipowm) [reply internal/provider/apgroup/resource_ap_group.go]: Confirmed — `gofmt -l` flagged this file: the two continuation lines of the `MarkdownDescription` string concatenation were over-indented (two tabs instead of the gofmt-expected single tab). Applied `gofmt -w internal/provider/apgroup/resource_ap_group.go`; the only diff is the two continuation lines re-indented from tab+tab to a single tab under `MarkdownDescription`. `gofmt -l internal/` is now clean and `make build` passes.

Note on verification: `make build` and gofmt pass. The other two checks could not be run to completion in this sandbox for reasons unrelated to this whitespace-only change — Docker is unavailable, so `make testacc` cannot start the controller, and the installed golangci-lint (v2.12.2) rejects the repo's v1-schema `.golangci.yml` (`unsupported version of the configuration`), a pre-existing environment mismatch. The linter the config enables is `gofmt`, which this change satisfies.

File: internal/provider/acctest/data_ap_group_test.go [hunk 13]
```diff
@@ -1,24 +1,51 @@
...
 import (
+	"fmt"
 	"github.com/hashicorp/terraform-plugin-testing/helper/resource"
 	"testing"
 )
 
+const (
+	testAPGroupDatasourceName = "data.unifi_ap_group.test"
+	defaultAPGroupName        = "All APs"
+)
```
[the remark this answers is not in the document]
Comment (filipowm) [reply internal/provider/acctest/data_ap_group_test.go:12]: Thanks for the catch. We're deliberately asserting the literal default name here: these acceptance tests run only against this project's Dockerized UniFi controller (started by the test harness in `acctest/provider_test.go`), where the built-in group is reliably `All APs` — it isn't user-configurable in that environment, so the assertion is stable rather than brittle.

Your broader point — that the data source should be pinned by its actual contract (`id` present, site resolved) and not just `name` — is well taken, and we're tracking that as a separate test-hardening item to add `TestCheckResourceAttrSet(..., "id")` plus a not-found case. Leaving this thread's specific change out for now.

File: internal/provider/acctest/resource_ap_group_test.go [hunk 14]
```diff
@@ -0,0 +1,175 @@
...
+	"strings"
+	"testing"
+)
+
+const testAPGroupResourceName = "unifi_ap_group.test"
+
+// FIXME: This test is currently skipped because the test environment does not support AP group adoption.
+func TestAccAPGroup_basic(t *testing.T) {
+	t.Skip("Skipping test due to lack of support for AP group adoption in test environment")
+	rName := acctest.RandomWithPrefix("tfacc-apgroup")
```
[the remark this answers is not in the document]
Comment (filipowm) [reply internal/provider/acctest/resource_ap_group_test.go:21]: Agreed that disabled lifecycle coverage is a real gap, but the skip is intentional and the suggested remedy doesn't address the root cause. `TestAccAPGroup_basic`/`_update` are `t.Skip`-ped because the Dockerized controller used by the test harness cannot adopt access points, so there are no real APs whose MACs an AP group can reference — gating the skip behind an env flag wouldn't create that capability; it would just turn a skip into a failure in the only lane we have.

To cover the conversion logic that the skipped acceptance tests were meant to exercise, the right move is a controller-free unit test over `APGroupModel.AsUnifiModel`/`Merge`; we're tracking that separately. Keeping this thread open as a known limitation rather than applying the env-flag change.

File: internal/provider/validators/regex.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (filipowm) [reply internal/provider/validators/regex.go:19]: The underlying observation is correct — the current pattern does accept mixed separators (I verified `aa:bb-cc:dd-ee:ff` matches), which is sloppy. But the proposed committable suggestion can't be merged as-is: it relies on a backreference (`\2`) to force a single separator, and Go's `regexp` package is RE2-based and has no backreference support. `regexp.MustCompile` on that pattern panics at package init:

```
error parsing regexp: invalid escape sequence: `\2`
```

Since this is a package-level `MustCompile`, that would crash the provider on load, not just fail validation. A RE2-safe way to forbid mixed separators is an alternation, e.g.:

```go
MacRegex = regexp.MustCompile(`^(([0-9A-Fa-f]{2}:){5}|([0-9A-Fa-f]{2}-){5})[0-9A-Fa-f]{2}$`)
```

I'm holding off on changing this here: MAC handling for the new resource is still under active design discussion (canonicalization at plan time vs. tightening the validator), and that decision should drive whether we tighten this shared regex (it's used by other resources too) or normalize separately. Leaving the thread open rather than applying a non-compiling fix.

File: internal/provider/apgroup/resource_ap_group.go (hunk 1 above)
Comment (filipowm) [reply internal/provider/apgroup/resource_ap_group.go]: Fixed in 8b90d7a. Added a set plan modifier (`apgroup.NormalizeMAC`) that rewrites every `device_macs` element to canonical lowercase, colon-separated form via `utils.CleanMAC` at plan time — so the planned value matches the controller's canonical response — mirroring the legacy `utils.MacDiffSuppressFunc`. `AsUnifiModel` also normalizes defensively. This removes the "inconsistent result after apply" failure and the perpetual diff.

Verification: `go build ./...`, `go vet`, and `gofmt` are clean (go 1.24). `make testacc` was not run — no Docker available locally; the skipped `TestAccAPGroup_update` now feeds an uppercase MAC and asserts the normalized lowercase result, so it will exercise this once AP-group adoption is supported in the harness.

File: internal/provider/apgroup/resource_ap_group.go (hunk 2 above)
Comment (filipowm) [reply internal/provider/apgroup/resource_ap_group.go]: Fixed in 8b90d7a (committed together with F-001 since both touch `device_macs`). `device_macs` is now `schema.SetAttribute`/`types.Set` with `setvalidator`, `Merge` uses `types.SetValueFrom`, `AsUnifiModel` reads via `ElementsAs`, and a `TestCheckSetResourceAttr` helper was added for order-insensitive assertions. Membership is now compared order-insensitively, eliminating reorder diffs.

File: internal/provider/validators/regex.go (hunk 5 above)
Comment (filipowm) [reply internal/provider/validators/regex.go:19]: Fixed in ee79e5f. Added `internal/provider/validators/regex_test.go` with a table-driven test over `MacRegex` (and the `Mac` validator) covering accepted colon/dash forms, rejected garbage/length/format/non-hex/whitespace cases, and an explicit decision: mixed-separator MACs (e.g. `00-11:22:33-44:55`) are **accepted**, because they are normalized to canonical form downstream by the `device_macs` plan modifier (F-001). Tests pass (`go test ./internal/provider/validators/...`).

File: internal/provider/apgroup/resource_ap_group.go (hunk 7 above)
Comment (filipowm) [reply internal/provider/apgroup/resource_ap_group.go:109]: Fixed in a33af0b. Added `examples/resources/unifi_ap_group/resource.tf` and `import.sh` (site:id pattern, matching `unifi_firewall_zone`), then regenerated docs with `go generate ./tools/` — producing `docs/resources/ap_group.md` and refreshing `docs/data-sources/ap_group.md`. (An unrelated `firewall_zone` data-source doc drift produced by the same run was reverted to keep this focused.)
